crates/authifier/src/config/sso.rs

```diff
@@ -0,0 +1,122 @@
+use std::{
+    borrow::Borrow,
+    collections::{HashMap, HashSet},
+    hash::{Hash, Hasher},
+    ops::Deref,
+};
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "lowercase", tag = "type")]
+pub enum Endpoints {
+    Discoverable,
+    Manual {
+        authorization: String,
+        token: String,
+        userinfo: String,
+    },
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "lowercase", tag = "type")]
+pub enum Credentials {
+    None {
+        client_id: String,
+    },
+    Basic {
+        client_id: String,
+        client_secret: String,
+    },
+    Post {
+        client_id: String,
+        client_secret: String,
+    },
+}
+
+impl Credentials {
+    pub fn client_id(&self) -> &str {
+        match self {
+            Credentials::None { client_id }
+            | Credentials::Basic { client_id, .. }
+            | Credentials::Post { client_id, .. } => client_id,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
+#[serde(rename_all = "lowercase")]
+pub enum Claim {
+    Id,
+    Username,
+    Picture,
+    Email,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct IdProvider {
+    pub id: String,
+
+    pub issuer: String,
+    pub name: Option<String>,
+    pub icon: Option<String>,
+
+    pub scopes: Vec<String>,
+    pub endpoints: Endpoints,
+    pub credentials: Credentials,
+    pub claims: HashMap<Claim, String>,
```

Comment:
This allows flexibility for claims, claims are a hashmap of information about the user, returned in the token or userinfo response i.e. ("preferred_email" -> "example@gmail.com", etc.) and this will allow cases where we wanna point out that "preferred_email" corresponds to the `Email` claim in case of unconventional keys.